software/contrib/gate_phaser.py

```diff
@@ -0,0 +1,211 @@
+        while True:
+            self.getKnobValues()
+            if self.screenRefreshNeeded:
+                self.updateScreen()
+            
+            # Cycle through outputs turning gates on and off as needed
+            # When a gate is triggered it's next on and off time is calculated
+            self.currentTimeStampMs = ticks_ms()
+            for n in range(len(cvs)):
+                if self.currentTimeStampMs >= self.gateOffTimes[n] and self.gateStates[n]:
+                    cvs[n].off()
+                    self.gateStates[n] = False
+                elif self.currentTimeStampMs >= self.gateOnTimes[n] and not self.gateStates[n]:
+                    cvs[n].on()
+                    self.gateStates[n] = True
+                    # When will the gate need to turn off?
+                    self.gateOffTimes[n] = self.currentTimeStampMs + GATE_LENGTH_MS
+                    # When will the next gate need to fire?
+                    self.gateOnTimes[n] = self.currentTimeStampMs + self.gateDelays[n] + self.masterGateIntervalMs
+
+            # Save state
+            if self.pendingSaveState and ticks_diff(ticks_ms(), self.lastSaveState) >= MIN_MS_BETWEEN_SAVES:
+                self.saveState()
+                self.pendingSaveState = False
+
+    def updateScreen(self):
+        """Update the screen only if something has changed. oled.show() hogs the processor and causes latency."""
+
+        # Clear screen
+        oled.fill(0)
+
+        oled.text("Cycle", 5, 0, 1)
+        oled.text(str(self.masterGateIntervalMs), 5, 10, 1)
+        oled.text("Delay", 80, 0, 1)
+        oled.text(str(self.slaveGateIntervalMs), 80, 10, 1)
+        oled.text(self.intervalsStr[:-1], 0, 22, 1)
+        oled.text('x' + str(self.gateDelayControlOptions[self.selectedGateControlMultiplier]), 104, 22, 1)
+
+        oled.show()
+        self.screenRefreshNeeded = False
+
+    def saveState(self):
+        """Save working vars to a save state file"""
+        self.state = {
+            "masterGateIntervalMs": self.masterGateIntervalMs,
+            "slaveGateIntervalMs": self.slaveGateIntervalMs,
+            "selectedGateDelayMultiple": self.selectedGateDelayMultiple,
+            "selectedGateControlMultiplier": self.selectedGateControlMultiplier,
+        }
+        self.save_state_json(self.state)
+        self.lastSaveState = ticks_ms()
+
+    def loadState(self):
+        """Load a previously saved state, or initialize working vars, then save"""
+        self.state = self.load_state_json()
+        self.masterGateIntervalMs = self.state.get("masterGateIntervalMs", 1000)
+        self.slaveGateIntervalMs = self.state.get("slaveGateIntervalMs", 100)
+        self.selectedGateDelayMultiple = self.state.get("selectedGateDelayMultiple", 0)
+        self.selectedGateControlMultiplier = self.state.get("selectedGateControlMultiplier", 0)
+
+if __name__ == "__main__":
+    dm = GatePhaser()
+    dm.main()
```

Comment:
This can be simplified to just
```
GatePhaser().main()
```

In general we try to keep the `__main__` section formatted like this across all of the contrib scripts.